### PR TITLE
feat(serverless): Serverless graph vertices

### PR DESCRIPTION
**Pull request by omriyoffe-panw**
# User description
**By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.**

[//]: # "
    # PR Title
    We use the title to create changelog automatically and therefore only allow specific prefixes
    - break:    to indicate a breaking change, this supersedes any of the other types
    - feat:     to indicate new features or checks
    - fix:      to indicate a bugfix or handling of edge cases of existing checks
    - docs:     to indicate an update to our documentation
    - chore:    to indicate adjustments to workflow files or dependency updates
    - platform: to indicate a change needed for the platform
    Each prefix should be accompanied by a scope that specifies the targeted framework. If uncertain, use 'general'.
    #    
    Allowed prefixs:
    ansible|argo|arm|azure|bicep|bitbucket|circleci|cloudformation|dockerfile|github|gha|gitlab|helm|kubernetes|kustomize|openapi|sast|sca|secrets|serverless|terraform|general|graph|terraform_plan|terraform_json
    #
    ex.
    feat(terraform): add CKV_AWS_123 to ensure that VPC Endpoint Service is configured for Manual Acceptance
"

## Description

*Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.*

Fixes # (issue)

### Description
*Include a description of what makes it a violation and any relevant external links.*

### Fix
*How does someone fix the issue in code and/or in runtime?*

## Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my feature, policy, or fix is effective and works
- [ ] New and existing tests pass locally with my changes

---

# Generated description

Below is a concise technical summary of the changes proposed in this PR:
<p>This pull request introduces enhancements to the serverless graph implementation, including new utility functions and test cases for improved functionality and coverage.</p>
    <table><tr><th>Topic</th><th>Details</th><tr><td><a href=https://baz.co/changes/bridgecrewio/checkov/6894?tool=ast&topic=Graph+implementation>Graph implementation</a>
        </td><td>Implements a graph-based approach for parsing and analyzing serverless configurations<details><summary>Modified files (7)</summary><ul><li>checkov/serverless/graph_builder/local_graph.py</li>
<li>checkov/serverless/runner.py</li>
<li>tests/serverless/graph_builder/test_local_graph.py</li>
<li>tests/serverless/graph_builder/resources/serverless.yaml</li>
<li>checkov/serverless/graph_builder/graph_to_definitions.py</li>
<li>tests/serverless/graph_builder/resources/serverless.yml</li>
<li>checkov/serverless/graph_builder/definition_context.py</li></ul></details><details><summary>Latest Contributors(2)</summary><table><tr><th>User</th><th>Commit</th><th>Date</th></tr><tr><td>omriyoffe-panw</td><td>feat-serverless-Server...</td><td>December 04, 2024</td></tr>
<tr><td>gruebel</td><td>chore-enable-mypy-for-...</td><td>November 27, 2023</td></tr></table></details></td></tr>
<tr><td><a href=https://baz.co/changes/bridgecrewio/checkov/6894?tool=ast&topic=Serverless+utils>Serverless utils</a>
        </td><td>Adds utility functions and enums to support serverless elements processing<details><summary>Modified files (2)</summary><ul><li>checkov/serverless/graph_builder/graph_components/block_types.py</li>
<li>checkov/serverless/utils.py</li></ul></details><details><summary>Latest Contributors(1)</summary><table><tr><th>User</th><th>Commit</th><th>Date</th></tr><tr><td>omriyoffe-panw</td><td>feat-serverless-Server...</td><td>December 04, 2024</td></tr></table></details></td></tr>
<tr><td><a href=https://baz.co/changes/bridgecrewio/checkov/6894?tool=ast&topic=Graph+tests>Graph tests</a>
        </td><td>Provides test coverage for the local graph implementation<details><summary>Modified files (2)</summary><ul><li>tests/serverless/graph_builder/__init__.py</li>
<li>tests/serverless/graph_builder/test_local_graph.py</li></ul></details><details><summary>Latest Contributors(0)</summary><table><tr><th>User</th><th>Commit</th><th>Date</th></tr></table></details></td></tr></table>This pull request is reviewed by Baz. Join @omriyoffe-panw and the rest of your team on <a href=https://baz.co/login>(Baz)</a>.

### CODE REVIEW

File: checkov/serverless/runner.py [hunk 1]
```diff
@@ -111,6 +113,18 @@ def run(
             # Filter out empty files that have not been parsed successfully
             self.definitions = {k: v for k, v in definitions.items() if v}
             self.definitions_raw = {k: v for k, v in definitions_raw.items() if k in definitions.keys()}
+            self.context = build_definitions_context(definitions=self.definitions, definitions_raw=self.definitions_raw)
+
+            if self.graph_registry and self.graph_manager:
```
Comment (bo156) [inline checkov/serverless/runner.py:118]: isn't this condition always True after your PR?

**Review by bo156 — APPROVED** [review]

File: checkov/serverless/runner.py (hunk 1 above)
Comment (omriyoffe-panw) [reply checkov/serverless/runner.py:118]: Yes you are right, probably was left over from an old implementation. I will remove it

**Review by SteveVaknin — APPROVED** [review]
Nice

File: checkov/serverless/runner.py [hunk 2]
```diff
@@ -140,34 +153,34 @@ def complete_python_checks(self,
...
+        entity = EntityDetails(sls_context_parser.provider_type, sls_file_data)
+        results = complete_registry.scan(sls_file, entity, skipped_checks, runner_filter)
+        tags = cfn_utils.get_resource_tags(entity, complete_registry)  # type:ignore[arg-type]
+        if results:
+            for check, check_result in results.items():
+                record = Record(check_id=check.id, check_name=check.name, check_result=check_result,
+                                code_block=[],  # Don't show, could be large
+                                file_path=self.extract_file_path_from_abs_path(Path(sls_file)),
+                                file_line_range=entity_lines_range,
+                                resource="complete",  # Weird, not sure what to put where
```
[the remark this answers is not in the document]
Comment (omriyoffe-panw) [reply checkov/serverless/runner.py:169]: Most of the runner's code existed before my change and I only refactored it. This comment was part of the original implementation. Not sure what he meant by that but didn't want to change to much

**Pull request merged**